input/pagecontent/smart-web-messaging.md

```diff
@@ -388,15 +410,255 @@ targetWindow.postMessage({
       // additional details as needed
     }
   }
-}, targetOrigin);
+}, ehrOrigin);
+```
+
+The EHR then performs the requested operation, creating an object in an internal
+scratchpad and assigning it id `123` before finally responding with:
+
+```js
+// From EHR -> app
+appWindow.postMessage({
+  "messageId": "<some new uid>",
+  "responseToMessageId": "<corresponding request messageId>",
+  "payload": {
+    "status": "201 CREATED",
+    "location": "ServiceRequest/123"
+  }
+}, appOrigin);
+```
+
+##### `scratchpad.read`
+
+###### Request `payload`
+
+| Property              | Optionality | Type   | Description |
+| --------------------- | ----------- | ------ | ----------- |
+| `location`            | OPTIONAL    | string | Takes the form `ResourceType/Id` when provided.  |
+{:.grid}
+
+###### Response `payload`
+
+| Property              | Optionality | Type   | Description |
+| --------------------- | ----------- | ------ | ----------- |
+| `status`              | REQUIRED    | string | An HTTP response code (i.e. "200 OK" or "404 NOT FOUND"). |
+| `resources`           | OPTIONAL    | object | [FHIR Bundle] containing zero or more resources matching the requested `location`. |
+| `outcome`             | OPTIONAL    | object | [FHIR OperationOutcome] resulting from the message action. |
+{:.grid}
+
+All resources included in the response `resources` bundle SHALL contain a
+`location` value which matches the value specified in the request `location`
+field; however, if no value was provided in the request, the response bundle
+SHALL contain the *full* scratchpad contents.
+
+If a `location` was specified in a request but not found in the scratchpad,
+or if there are no entries in the scratchpad, the `resources` property may be
+omitted from the response payload.
+
+###### Examples
+
+---
+
+This example shows how an app might request the contents of a single resource
+from the scratchpad, `ServiceRequest/123`.
+
+```js
+// From app -> EHR
+ehrWindow.postMessage({
+  "messageId": "<some new uid>",
+  "messagingHandle": "<smart_web_messaging_handle> from SMART launch context",
+  "messageType": "scratchpad.read",
+  "payload": {
+    "location": "ServiceRequest/123"
+  }
+}, ehrOrigin);
+```
+
+Assuming the resource exists, the EHR could return it to the app like this:
+
+```js
+// From EHR -> app
+appWindow.postMessage({
+  "messageId": "<some new uid>",
+  "responseToMessageId": "<corresponding request messageId>",
+  "payload": {
+    "status": "200 OK",
+    "resources": {
+      "id": "<some new uid>",
+      "resourceType": "Bundle",
+      "type": "collection",
+      "entry": [
+        {
+          "fullUrl": "ServiceRequest/123",
```

Comment:
I don't think we can leave `fullUrl` as a relative URL like this -- per http://build.fhir.org/bundle-definitions.html#Bundle.entry.fullUrl I'd either 1) omit fullUrl, or 2) just provide an array of Bundle.entry.responses instead of a full Bundle here, to avoid the issue.

Reply:
I think the [regex](https://build.fhir.org/references.html#regex) that validates the url *does* allow relative URLs, but I think there is also a note suggesting that it should refer to a real resource if specified (which this wouldn't).  But I'm not sure I'm interpreting the comments right.

> Whether or not the reference is to a FHIR RESTful server, the reference SHALL point to a Resource as defined by this specification.

And:
>References SHALL be a reference to an actual FHIR resource, and SHALL be resolvable (given that access control works, there is no temporary unavailability, etc.). Resolution can be either by retrieval from the URL, or, where applicable by resource type, by treating an absolute reference as a canonical URL (see below) and looking it up in a local registry/repository

The problem I was trying to solve is - how to provide the `scratchpad.location` in the returned resource.  If I return a list of Bundle.entry.responses - I think I can set the [`location`](https://build.fhir.org/bundle-definitions.html#Bundle.entry.response.location) and the resource becomes the [`outcome`](https://build.fhir.org/bundle-definitions.html#Bundle.entry.response.outcome).

Reply:
Uh, I don't know what I was thinking suggesting that the scratchpad resource be inserted in the `Bundle.entry.response.outcome` field.  That's reserved for `OperationOutcome` objects, not resources.

Part of the problem is that if we omit `fullUrl`, the bundle will no longer validate.  The `fullUrl` attribute is required for bundles that are not transactions or batches.

If that's OK, we could replace the `fullUrl` with a new `link` in each `Bundle.entry` that might look like this:
```json
{
  "entry": [
    {
      "link": {
        "working-copy": "Patient/123"
      },
...
    }
  ]
}
```

I puzzled over the different link types specified here:
https://www.iana.org/assignments/link-relations/link-relations.xhtml#link-relations-1

I guess I could see an argument for `self`, `working-copy`, `item`, or `current`.

Reply:
Rather than add more complexity definiting a new type of link that only makes sense for scratchpad objects - I've opted to simply require that Resources in the scratchpad include `resourceType` and `id` attributes, and that the scratchpad use those values to construct the `scratchpad.location` handle.